### PR TITLE
Fold a bunch of CoreCLR's abandoned mutex tests into a parameterized test

### PR DESCRIPTION
**Pull request by kouvel**
- There are a large number of tests in CoreCLR with a lot of redundancy, timing issues, and which take unnecessarily long to run
- Folded a bunch of them into the existing parameterized test and improved it to cover more cases to replace the tests in CoreCLR that will be removed
- Coverage is not 100% equivalent, but I believe this sufficiently covers the necessary aspects of the tests in CoreCLR that will be removed

### CODE REVIEW

**Comment by kouvel** [comment]
Current test failures are due to https://github.com/dotnet/coreclr/pull/25452, once that change makes it into corefx they should pass

File: src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs [hunk 1]
```diff
@@ -29,6 +29,7 @@ public static partial class AssertExtensions
         public static void ThrowsIf<T>(bool condition, System.Action action) where T : System.Exception { }
         public static T Throws<T>(System.Action action) where T : System.Exception { throw null; }
         public static void Throws<T>(System.Action action, string message) where T : System.Exception { }
+        public static TException Throws<TException, TResult>(Func<TResult> func) where TException : Exception { throw null; }
```
Comment (stephentoub) [inline src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs]: Why is this needed, rather than just using `Assert.Throws`?

File: src/System.Threading/tests/MutexTests.cs [hunk 2]
```diff
@@ -165,45 +172,182 @@ private static IEnumerable<string> GetNamePrefixes()
         public static IEnumerable<object[]> AbandonExisting_MemberData()
         {
             var nameGuidStr = Guid.NewGuid().ToString("N");
-            for (int waitType = 0; waitType < 2; ++waitType) // 0 == WaitOne, 1 == WaitAny
+            for (var waitType = WaitHandleWaitType.WaitOne; waitType <= WaitHandleWaitType.WaitAll; ++waitType)
```
Comment (stephentoub) [inline src/System.Threading/tests/MutexTests.cs]: `foreach (WaitHandleWaitType waitType in Enum.GetValues(typeof(WaitHandleWaitType)))`?

**Comment by stephentoub** [comment]
The Unix legs are all failing on:
```
AbandonExisting(name: null, waitType: WaitAll, waitCount: 3, notAbandonedWaitIndex: 2, isNotAbandonedWaitObjectSigna)
```
with, e.g.
```
Ubuntu.1804.Amd64.Open-x64-Debug
Get Repro environment
Unhandled Exception of Type System.AggregateException
Message :
System.AggregateException : One or more errors occurred. (Result: True (Assert.Throws() Failure
Expected: typeof(System.Threading.AbandonedMutexException)
Actual:   (No exception was thrown)))
---- System.AggregateException : Result: True (Assert.Throws() Failure
Expected: typeof(System.Threading.AbandonedMutexException)
Actual:   (No exception was thrown))
-------- Assert.Throws() Failure
Expected: typeof(System.Threading.AbandonedMutexException)
Actual:   (No exception was thrown)
Stack Trace :
   at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__1() in /_/src/Common/tests/System/Threading/ThreadTestHelpers.cs:line 52
   at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__2() in /_/src/Common/tests/System/Threading/ThreadTestHelpers.cs:line 57
   at System.Threading.Tests.ThreadTestHelpers.RunTestInBackgroundThread(Action test) in /_/src/Common/tests/System/Threading/ThreadTestHelpers.cs:line 108
   at System.Threading.Tests.MutexTests.AbandonExisting(String name, WaitHandleWaitType waitType, Int32 waitCount, Int32 notAbandonedWaitIndex, Boolean isNotAbandonedWaitObjectSignaled, Boolean abandonDuringWait) in /_/src/System.Threading/tests/MutexTests.cs:line 241
----- Inner Stack Trace -----
   at System.AssertExtensions.Throws[TException,TResult](Func`1 func) in /_/src/CoreFx.Private.TestUtilities/src/System/AssertExtensions.cs:line 108
   at System.Threading.Tests.MutexTests.<>c__DisplayClass13_0.<AbandonExisting>b__0() in /_/src/System.Threading/tests/MutexTests.cs:line 336
   at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__0() in /_/src/Common/tests/System/Threading/ThreadTestHelpers.cs:line 36
----- Inner Stack Trace -----
   at System.AssertExtensions.Throws[TException,TResult](Func`1 func) in /_/src/CoreFx.Private.TestUtilities/src/System/AssertExtensions.cs:line 85
```

File: src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs (hunk 1 above)
Comment (kouvel) [reply src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs]: I wanted to log the return value when an exception is not thrown

**Comment by kouvel** [comment]
Looks like https://github.com/dotnet/coreclr/pull/25452 still hasn't made it into corefx, tests should pass after that

**Comment by stephentoub** [comment]
> Looks like dotnet/coreclr#25452 still hasn't made it into corefx, tests should pass after that

Are you sure? We've definitely consumed coreclr updates on the past three weeks.

**Comment by kouvel** [comment]
Hmm not sure what was wrong. Maybe I needed to rebase to a more recent commit, I synced again and it's passing locally for me. I have pushed the rebased version, trying again.
